### PR TITLE
[Bounds Widening] Compute Gen and Kill sets

### PR DESCRIPTION
**Pull request by mgrang**

### CODE REVIEW

File: clang/include/clang/Sema/BoundsWideningAnalysis.h [hunk 1]
```diff
@@ -56,14 +67,139 @@ namespace clang {
...
+    void GetVarsInExpr(Expr *E, VarSetTy &VarsInExpr) const;
+
+    // Invoke IgnoreValuePreservingOperations to strip off casts.
+    // @param[in] E is the expression whose casts must be stripped.
+    // @return E with casts stripped off.
+    Expr *IgnoreCasts(const Expr *E) const;
+
+    // We do not want to run dataflow analysis on null blocks or the exit
+    // block. So we skip them.
+    // @param[in] B is the block which may need to the skipped from dataflow
```
Comment (kkjeer) [inline clang/include/clang/Sema/BoundsWideningAnalysis.h:180]: `the skipped` => `be skipped`

File: clang/include/clang/Sema/BoundsWideningAnalysis.h [hunk 2]
```diff
@@ -56,14 +67,139 @@ namespace clang {
...
+    // Check if V is an _Nt_array_ptr or an _Nt_checked array.
+    // @param[in] V is a VarDecl.
+    // @return Whether V is an _Nt_array_ptr or an _Nt_checked array.
+    bool IsNtArrayType(const VarDecl *V) const;
+
+    // Compute the set difference of sets A and B.
+    // @param[in] A is a set.
+    // @param[in] B is a set.
+    // @return The set difference of sets A and B.
+    template<class T, class U> T Difference(T &A, U &B) const;
```
Comment (kkjeer) [inline clang/include/clang/Sema/BoundsWideningAnalysis.h:194]: Would it make sense to put `Difference` and `Set` in ExprUtils.h or another utility file?

File: clang/lib/Sema/BoundsWideningAnalysis.cpp [hunk 3]
```diff
@@ -14,4 +14,396 @@
...
+    return ExprCreatorUtil::CreateBinaryOperator(SemaRef, AE->getBase(),
+                                                 AE->getIdx(),
+                                                 BinaryOperatorKind::BO_Add);
+  }
+  return nullptr;
+}
+
+void BoundsWideningAnalysis::GetVarsInExpr(Expr *E,
+                                           VarSetTy &VarsInExpr) const {
+  // Get the VarDecls of all variables occurring in an expression.
```
Comment (kkjeer) [inline clang/lib/Sema/BoundsWideningAnalysis.cpp:343]: It seems like `GetVarsInExpr` returns a subset of all variables occurring in an expression. For example, in the expression `s->f + myFunction(x)`, `GetVarsInExpr` will return an empty set, not `{ s, x }`. Could this function include a comment stating that it only returns some of the variables that appear in `E`?

File: clang/lib/Sema/BoundsWideningAnalysis.cpp [hunk 4]
```diff
@@ -14,4 +14,396 @@
...
+      continue;
+
+    for (const VarDecl *V : It->second)
+      EB->Kill[S].insert(V);
+  }
+}
+
+void BoundsWideningAnalysis::GetModifiedVars(const Stmt *S,
+                                             VarSetTy &ModifiedVars) {
+  // Get all variables modfied by statement S or statements nested in S.
```
Comment (kkjeer) [inline clang/lib/Sema/BoundsWideningAnalysis.cpp:229]: `modfied` => `modified`

**Comment by mgrang** [comment]
Closing in favor of https://github.com/microsoft/checkedc-clang/pull/1050/commits/6530a05cddd919b0cef93046abef48afbdef274d

**Pull request closed without merging**